host-containers: make host containers inherit proxy env vars

etungsten opened the pull request:
<!--
Tips:
- Please read CONTRIBUTING.md to understand our process and our requests for PRs.
- Please file an issue before creating a PR so we can discuss the change and confirm it's not already being worked on.
-->

**Issue number:**
Fixes https://github.com/bottlerocket-os/bottlerocket/issues/1424


**Description of changes:**
```
Author: Erikson Tung <etung@amazon.com>
Date:   Tue Mar 30 14:39:00 2021 -0700

    host-containers: make host containers inherit proxy env vars
    
    host-containers systemd services now reads proxy env vars.
    host-ctr passes the proxy env vars to the host containers.
    host-containers are added to the list of affected services under
    `metadata.settings.network`
```

```
Author: Erikson Tung <etung@amazon.com>
Date:   Thu Apr 1 13:13:16 2021 -0700

    migration-helpers: new ReplaceMetadataListsMigration helper
    
    Adds a new migration helper for migrating setting's metadata list changes

```

```
Author: Erikson Tung <etung@amazon.com>
Date:   Thu Apr 1 13:24:39 2021 -0700

    migrations: add migration for list of network proxy affected-services
    
    Adds a new migration for migrating the 'affected-services' list for
    'settings.network' to include 'host-containers' on upgrades and restore
    old list values on downgrade.

```

**Testing done:**
Set up a simple http proxy 
Built AMI, launched instance with userdata specifying my proxy server:
```
[settings.network]
https-proxy="somewhere:9898"
no-proxy = [".local"]
```

Instance came up successfully, was able to ssh to admin container and sudo sheltie:
Both host-containers@control and host-containers@admin services ran fine:

I started an SSM session with the control host container and was able to verify the environment variables were successfully inherited:
```
[ssm-user@ /]$ echo $HTTPS_PROXY 
somewhere:9898
[ssm-user@ /]$ echo $NO_PROXY 
.local,localhost,127.0.0.1,abcdefghijklmnopqrstuv.us-west-2.eks.amazonaws.com,.cluster.local
[ssm-user@ /]$ 
```

- Migration helper: All unit tests pass

- Migration test: Build aws-k8s-1.19 AMI off v1.0.7 release and created a TUF repository with a aws-k8s-1.19 with my changes and migration with release version set to v1.0.8.

Launched v1.0.7 aws-k8s-1.19.
Forward migration (after updating to v1.0.8):
```
bash-5.0# cat /var/lib/bottlerocket/datastore/current/live/settings/network.affected-services 
["containerd","kubernetes","host-containerd","host-containers"]
```
Backwards migration (after downgrading to v1.0.7):
```
bash-5.0# cat /var/lib/bottlerocket/datastore/current/live/settings/network.affected-services 
["containerd","kubernetes","host-containerd"]
```



**Note that the default admin container's entry point is [`start_admin_sshd.sh`](https://github.com/bottlerocket-os/bottlerocket-admin-container/blob/79a71586e0d56c9bab07d4b0830d5ee3766ed228/start_admin_sshd.sh)  and at the end of the script it execs the sshd daemon and proxy environment variables won't be present in the sshd spawned shells**
This is being addressed in https://github.com/bottlerocket-os/bottlerocket-admin-container/pull/24

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

etungsten committed:
host-containers: make host containers inherit proxy env vars

host-containers systemd services now reads proxy env vars.
host-ctr passes the proxy env vars to the host containers.
host-containers are added to the list of affected services under
`metadata.settings.network` for all variants

bcressey commented:
> Note that the default admin container's entry point is start_admin_sshd.sh and at the end of the script it execs the sshd daemon and proxy environment variables won't be present in the sshd spawned shells

We'd need to fix in the admin container by writing these values into the default user environment files under `/etc`.

bcressey commented on sources/host-ctr/cmd/host-ctr/main.go:
```diff
@@ -213,6 +213,7 @@ func runCtr(containerdSocket string, namespace string, containerID string, sourc
 			oci.WithHostNamespace(runtimespec.NetworkNamespace),
 			oci.WithHostHostsFile,
 			oci.WithHostResolvconf,
+			oci.WithEnv([]string{"HTTPS_PROXY=" + os.Getenv("HTTPS_PROXY"), "NO_PROXY=" + os.Getenv("NO_PROXY")}),
```
Should we check whether `HTTPS_PROXY` is set before adding it to the environment? Applications might handle "set but empty" and "unset" differently.

Should we also export `https_proxy` and `no_proxy`, as we've done in `proxy.env`?

etungsten commented:
Push above addresses @bcressey 's [comment](https://github.com/bottlerocket-os/bottlerocket/pull/1432#discussion_r604511299):

Proxy env vars get inherited correctly:
```
[ssm-user@ip-192-168-9-239 /]$ env
TERM=xterm-256color
NO_PROXY=.local,localhost,127.0.0.1,abc.yl4.us-west-2.eks.amazonaws.com,.cluster.local
...
https_proxy=somewhere:9898
HTTPS_PROXY=somewhere:9898
...
IMAGE_VERSION=v0.4.2
no_proxy=.local,localhost,127.0.0.1,abc.yl4.us-west-2.eks.amazonaws.com,.cluster.local
...
_=/usr/bin/env

```

If I wipe proxy.env from the host and restart `host-containers@control`, the proxy env vars are no longer there as expected:
```
[ssm-user@ip-192-168-9-239 /]$ env
TERM=xterm-256color
LS_COLORS=rs=0:di=38;5;27:ln=38;5;51:mh=44;38;5;15:pi=40;38;5;11:so=38;5;13:do=38;5;5:bd=48;5;232;38;5;11:cd=48;5;232;38;5;3:or=48;5;232;38;5;9:mi=05;48;5;232;38;5;15:su=48;5;196;38;5;15:sg=48;5;11;38;5;16:ca=48;5;196;38;5;226:tw=48;5;10;38;5;16:ow=48;5;10;38;5;21:st=48;5;21;38;5;15:ex=38;5;34:*.tar=38;5;9:*.tgz=38;5;9:*.arc=38;5;9:*.arj=38;5;9:*.taz=38;5;9:*.lha=38;5;9:*.lz4=38;5;9:*.lzh=38;5;9:*.lzma=38;5;9:*.tlz=38;5;9:*.txz=38;5;9:*.tzo=38;5;9:*.t7z=38;5;9:*.zip=38;5;9:*.z=38;5;9:*.Z=38;5;9:*.dz=38;5;9:*.gz=38;5;9:*.lrz=38;5;9:*.lz=38;5;9:*.lzo=38;5;9:*.xz=38;5;9:*.bz2=38;5;9:*.bz=38;5;9:*.tbz=38;5;9:*.tbz2=38;5;9:*.tz=38;5;9:*.deb=38;5;9:*.rpm=38;5;9:*.jar=38;5;9:*.war=38;5;9:*.ear=38;5;9:*.sar=38;5;9:*.rar=38;5;9:*.alz=38;5;9:*.ace=38;5;9:*.zoo=38;5;9:*.cpio=38;5;9:*.7z=38;5;9:*.rz=38;5;9:*.cab=38;5;9:*.jpg=38;5;13:*.jpeg=38;5;13:*.gif=38;5;13:*.bmp=38;5;13:*.pbm=38;5;13:*.pgm=38;5;13:*.ppm=38;5;13:*.tga=38;5;13:*.xbm=38;5;13:*.xpm=38;5;13:*.tif=38;5;13:*.tiff=38;5;13:*.png=38;5;13:*.svg=38;5;13:*.svgz=38;5;13:*.mng=38;5;13:*.pcx=38;5;13:*.mov=38;5;13:*.mpg=38;5;13:*.mpeg=38;5;13:*.m2v=38;5;13:*.mkv=38;5;13:*.webm=38;5;13:*.ogm=38;5;13:*.mp4=38;5;13:*.m4v=38;5;13:*.mp4v=38;5;13:*.vob=38;5;13:*.qt=38;5;13:*.nuv=38;5;13:*.wmv=38;5;13:*.asf=38;5;13:*.rm=38;5;13:*.rmvb=38;5;13:*.flc=38;5;13:*.avi=38;5;13:*.fli=38;5;13:*.flv=38;5;13:*.gl=38;5;13:*.dl=38;5;13:*.xcf=38;5;13:*.xwd=38;5;13:*.yuv=38;5;13:*.cgm=38;5;13:*.emf=38;5;13:*.axv=38;5;13:*.anx=38;5;13:*.ogv=38;5;13:*.ogx=38;5;13:*.aac=38;5;45:*.au=38;5;45:*.flac=38;5;45:*.mid=38;5;45:*.midi=38;5;45:*.mka=38;5;45:*.mp3=38;5;45:*.mpc=38;5;45:*.ogg=38;5;45:*.ra=38;5;45:*.wav=38;5;45:*.axa=38;5;45:*.oga=38;5;45:*.spx=38;5;45:*.xspf=38;5;45:
ENV=/etc/skel/.bashrc
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
PWD=/
LANG=C.UTF-8
SHLVL=1
HOME=/home/ssm-user
IMAGE_VERSION=v0.4.2
ARCH=amd64
SSM_AGENT_VERSION=3.0.732.0
_=/usr/bin/env
```

etungsten commented:
Push above adds `host-containers` to the list of affected services under `metadata.settings.network` so when proxy settings are changed, `thar-be-settings` will kick off the `restart-commands` for `host-containers`

Note that until https://github.com/bottlerocket-os/bottlerocket/issues/1389 is resolved, host-containers aren't restarted until the user decides to cycle the host-container's `enabled` setting. For now, it serves to remind us to consider proxy setting changes when restarting host-containers (essentially we should restart all of them if proxy settings changes).

samuelkarp commented on sources/host-ctr/cmd/host-ctr/main.go:
```diff
@@ -606,6 +608,22 @@ func withPersistentStorage(containerID string) oci.SpecOpts {
+// withProxyEnv reads proxy environment variables and returns a spec option for passing said proxy environment variables
+func withProxyEnv() oci.SpecOpts {
+	noOp := func(_ context.Context, _ oci.Client, _ *containers.Container, s *runtimespec.Spec) error { return nil }
+	httpsProxy, httpsProxySet := os.LookupEnv("HTTPS_PROXY")
+	noProxy, noProxySet := os.LookupEnv("NO_PROXY")
+	withHttpsProxy := noOp
+	withNoProxy := noOp
+	if httpsProxySet {
+		withHttpsProxy = oci.WithEnv([]string{"HTTPS_PROXY=" + httpsProxy, "https_proxy=" + httpsProxy})
+	}
+	if noProxySet {
+		withNoProxy = oci.WithEnv([]string{"NO_PROXY=" + noProxy, "no_proxy=" + noProxy})
+	}
+	return oci.Compose(withHttpsProxy, withNoProxy)
+}
```
Do we need to set empty environment variables?  If we can skip that, `os.Getenv` ends up being a bit simpler.  And we can rewrite this in a slightly simpler way too.

```suggestion
	withHttpsProxy := noOp
	if val := os.Getenv("HTTPS_PROXY"); val != "" {
		withHttpsProxy = oci.WithEnv([]string{
			"HTTPS_PROXY=" + val,
			"https_proxy=" + val,
		})
	}
	withNoProxy := noOp
	if val := os.Getenv("NO_PROXY"); val != "" {
		withNoProxy = oci.WithEnv([]string{
			"NO_PROXY=" + val,
			"no_proxy=" + val,
		})
	}
	return oci.Compose(withHttpsProxy, withNoProxy)
}
```

etungsten replied:
I think we want to differentiate between the variables being unset and being null just in case there are applications that care about that distinction. (See https://github.com/bottlerocket-os/bottlerocket/pull/1432#discussion_r604511299)

samuelkarp replied:
That comment is about "set but empty" environment variables.  The approach I outlined above avoids that (checking for `val != ""`) and also avoids propagating "set but empty" environment variables into the host containers.  If we think "set but empty" variables _should_ be propagated, then we need `os.LookupEnv`.

etungsten replied:
> If we think "set but empty" variables should be propagated, then we need os.LookupEnv.

I understood the ask as to not propagate the environment variable if it was unset in the first place, but propagate it if it was set regardless of whether it's empty or not, thus the `os.LookupEnv`.

Can you please clarify, @bcressey?

webern replied:
It seems 'safer' to mirror the vars exactly (e.g. set-but-empty if that's what we find) rather than to assume we know better.

samuelkarp reviewed (approved):
Looks good, but a minor simplification suggestion

tjkirch commented on sources/models/shared-defaults/defaults.toml:
```diff
@@ -74,7 +74,7 @@ path = "/etc/network/proxy.env"
 
 [metadata.settings.network]
-affected-services = ["containerd", "host-containerd"]
+affected-services = ["containerd", "host-containerd", "host-containers"]
```
I think we need a migration for this.

etungsten commented:
Taking this PR back into draft, as I need to create a new migration helper that can replace metadata lists.

etungsten committed:
migration-helpers: new ReplaceMetadataListsMigration helper

Adds a new migration helper for migrating setting's metadata list changes

etungsten committed:
migrations: add migration for list of network proxy affected-services

Adds a new migration for migrating the 'affected-services' list for
'settings.network' to include 'host-containers' on upgrades and restore
old list values on downgrade.

etungsten commented:
Push above does the following:

- Adds "host-containers" to the list of settings.network's `affected-services` for all existing variants' defaults.
- Adds a new migration helper ReplaceMetadataListsMigration for migrating metadata list values
- Adds a new migration migrating the 'affected-services' list for 'settings.network' to include 'host-containers' on upgrades and restore old list values on downgrade.

Tested the migration with a live instance and TUF repository, and the migration works as expected. Please refer to the PR description for testing details.

etungsten commented:
Push above adds a missing new line at the end of a file and fixes a commit message.

etungsten commented:
Push above rebases onto develop and fixes conflicts.

tjkirch commented on Release.toml:
```diff
@@ -33,4 +33,5 @@ version = "1.0.7"
 "(1.0.7, 1.0.8)" = [
```
For lack of anywhere concrete to put this comment: the commit message has a typo, `migration heler`

tjkirch commented on sources/api/migration/migrations/v1.0.8/proxy-affect-host-containers/src/main.rs:
```diff
@@ -0,0 +1,64 @@
+            old_vals: &["containerd", "host-containerd"],
+            new_vals: &["containerd", "host-containerd", "host-containers"],
+        },
+        // For K8S variants
```
Hah, we finally have a variant-based difference in a migration, and it doesn't seem like it was too bad.  We'd run into trouble if two variants had the same settings before and we want them to migrate to different settings, but we have system info available and we'd be able to check as needed, so... doable, but not easy yet.  Hopefully we won't come to that position anyway.

webern replied:
I guess if we had the same `old_vals` in two different variants that needed to go to different `new_vals`, then this wouldn't have worked. But yeah.

tjkirch commented on sources/api/migration/migrations/v1.0.8/proxy-affect-host-containers/src/main.rs:
```diff
@@ -0,0 +1,64 @@
+                "host-containers",
+            ],
+        },
+        // For the aws-dev variant
```
```suggestion
        // For aws-dev and vmware-dev variants
```

tjkirch commented on sources/api/migration/migration-helpers/src/common_migrations.rs:
```diff
@@ -842,3 +842,306 @@ mod test_add_metadata {
+                    }
+                } else {
+                    println!(
+                        "Found no metadata '{}' to for setting '{}'",
```
```suggestion
                        "Found no metadata '{}' for setting '{}'",
```

etungsten commented:
Push above addresses @tjkirch's comments:

- Fixes typos

bcressey commented on sources/models/src/aws-dev/defaults.d/50-aws-dev.toml:
```diff
@@ -14,4 +14,4 @@ service-checks = ["apiserver", "chronyd", "containerd", "host-containerd", "dock
 
 # Network
 [metadata.settings.network]
-affected-services = ["containerd", "docker", "host-containerd"]
+affected-services = ["containerd", "docker", "host-containerd", "host-containers"]
```
We also need this in the `vmware-dev` variant.

etungsten commented:
Push above addresses @bcressey 's comment. 

- Added `host-containers` to the list of network affected-services for the vmware-dev variant.

zmrow reviewed (approved):
:national_park:
